File: web/src/dsl/runtime/NotePropertyHandler.ts
import type { Note } from '@/data/Note'
import type { DslValue, NoteVal } from './DslValue'
import { dateTimeVal, noteVal, stringVal, UNDEFINED } from './DslValue'
import type { Environment } from './Environment'
import { ExecutionException } from './ExecutionException'
import { MutationType } from './NoteMutation'

/**
 * Handles property access and assignment for NoteVal.
 */
export function getProperty(nv: NoteVal, property: string, env: Environment): DslValue {
  const note = nv.note
  switch (property) {
    case 'id': return stringVal(note.id)
    case 'path': return stringVal(note.path)
    case 'name': return stringVal(note.content.split('\n')[0] ?? '')
    case 'created': {
      if (!note.createdAt) throw new ExecutionException('Note has no created date')
      return dateTimeVal(note.createdAt.toDate().toISOString().slice(0, 19))
    }
    case 'modified': {
      if (!note.updatedAt) throw new ExecutionException('Note has no modified date')
      return dateTimeVal(note.updatedAt.toDate().toISOString().slice(0, 19))
    }
    case 'up': return getUp(note, 1, env)
    case 'root': return getRoot(nv, env)
    default: throw new ExecutionException(`Unknown property '${property}' on note`)
  }
}

export async function setProperty(nv: NoteVal, property: string, value: DslValue, env: Environment): Promise<void> {
  const note = nv.note
  const ops = env.getNoteOperations()
  if (!ops) throw new ExecutionException('Cannot modify note properties: note operations not available')

  switch (property) {
    case 'path': {
      if (value.kind !== 'StringVal') throw new ExecutionException('path must be a string')
      const updatedNote = await ops.updatePath(note.id, value.value)
      env.registerMutation({ noteId: note.id, updatedNote, mutationType: MutationType.PATH_CHANGED })
      break
    }
    case 'name': {
      if (value.kind !== 'StringVal') throw new ExecutionException('name must be a string')
      const freshNote = await ops.getNoteById(note.id)
      if (!freshNote) throw new ExecutionException(`Note not found: ${note.id}`)
      const lines = freshNote.content.split('\n')
      const newContent = lines.length <= 1 ? value.value : [value.value, ...lines.slice(1)].join('\n')
      const updatedNote = await ops.updateContent(note.id, newContent)
      env.registerMutation({ noteId: note.id, updatedNote, mutationType: MutationType.CONTENT_CHANGED })
      break
    }
    case 'id':
    case 'created':
    case 'modified':
      throw new ExecutionException(`Cannot set read-only property '${property}' on note`)
    default:
      throw new ExecutionException(`Unknown property '${property}' on note`)
  }
}

export function getUp(note: Note, levels: number, env: Environment): DslValue {
  if (levels === 0) return noteVal(note)
  const parent = env.getParentNote(note)
  if (!parent) return UNDEFINED
  if (levels === 1) return noteVal(parent)
  return getUp(parent, levels - 1, env)
}

function getRoot(nv: NoteVal, env: Environment): NoteVal {
  const parent = env.getParentNote(nv.note)
  if (!parent) return nv
  return getRoot(noteVal(parent), env)
}

File: app/src/main/java/org/alkaline/taskbrain/dsl/runtime/NotePropertyHandler.kt
package org.alkaline.taskbrain.dsl.runtime

import org.alkaline.taskbrain.data.Note
import org.alkaline.taskbrain.dsl.runtime.values.DateTimeVal
import org.alkaline.taskbrain.dsl.runtime.values.DslValue
import org.alkaline.taskbrain.dsl.runtime.values.NoteVal
import org.alkaline.taskbrain.dsl.runtime.values.StringVal
import org.alkaline.taskbrain.dsl.runtime.values.UndefinedVal
import java.time.ZoneId

/**
 * Handles property access and assignment for NoteVal.
 *
 * Supported readable properties:
 * - `id`: String - Firebase ID
 * - `path`: String - Unique path identifier
 * - `name`: String - First line of content
 * - `created`: DateTime - Creation timestamp
 * - `modified`: DateTime - Last modification timestamp
 * - `up`: Note - Parent note (1 level up)
 * - `root`: Note - Root ancestor (top-level note with no parent)
 *
 * Writable properties:
 * - `path`: String - Unique path identifier
 * - `name`: String - First line of content (updates content)
 */
object NotePropertyHandler {

    /**
     * Get a property value from the note.
     */
    fun getProperty(noteVal: NoteVal, property: String, env: Environment): DslValue {
        val note = noteVal.note
        return when (property) {
            "id" -> StringVal(note.id)
            "path" -> StringVal(note.path)
            "name" -> StringVal(note.content.lines().firstOrNull() ?: "")
            "created" -> note.createdAt?.let {
                DateTimeVal(it.toDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime())
            } ?: throw ExecutionException("Note has no created date")
            "modified" -> note.updatedAt?.let {
                DateTimeVal(it.toDate().toInstant().atZone(ZoneId.systemDefault()).toLocalDateTime())
            } ?: throw ExecutionException("Note has no modified date")
            "up" -> getUp(note, 1, env)
            "root" -> getRoot(noteVal, env)
            else -> throw ExecutionException("Unknown property '$property' on note")
        }
    }

    /**
     * Set a property value on the note.
     */
    fun setProperty(noteVal: NoteVal, property: String, value: DslValue, env: Environment) {
        val note = noteVal.note
        val ops = env.getNoteOperations()
            ?: throw ExecutionException(
                "Cannot modify note properties: note operations not available"
            )

        when (property) {
            "path" -> {
                val newPath = (value as? StringVal)?.value
                    ?: throw ExecutionException("path must be a string")
                val updatedNote = kotlinx.coroutines.runBlocking {
                    ops.updatePath(note.id, newPath)
                }
                env.registerMutation(NoteMutation(note.id, updatedNote, MutationType.PATH_CHANGED))
            }
            "name" -> {
                val newName = (value as? StringVal)?.value
                    ?: throw ExecutionException("name must be a string")
                // Fetch fresh content from Firestore to avoid stale cache issues
                val freshNote = kotlinx.coroutines.runBlocking {
                    ops.getNoteById(note.id)
                } ?: throw ExecutionException("Note not found: ${note.id}")
                // Update the first line of content while preserving the rest
                val lines = freshNote.content.lines()
                val newContent = if (lines.isEmpty() || lines.size == 1) {
                    newName
                } else {
                    (listOf(newName) + lines.drop(1)).joinToString("\n")
                }
                val updatedNote = kotlinx.coroutines.runBlocking {
                    ops.updateContent(note.id, newContent)
                }
                env.registerMutation(NoteMutation(note.id, updatedNote, MutationType.CONTENT_CHANGED))
            }
            "id", "created", "modified" -> {
                throw ExecutionException("Cannot set read-only property '$property' on note")
            }
            else -> throw ExecutionException("Unknown property '$property' on note")
        }
    }

    /**
     * Navigate up the note hierarchy by the specified number of levels.
     * Returns UndefinedVal if the requested ancestor doesn't exist.
     */
    internal fun getUp(note: Note, levels: Int, env: Environment): DslValue {
        if (levels == 0) return NoteVal(note)
        val parent = env.getParentNote(note) ?: return UndefinedVal
        return if (levels == 1) {
            NoteVal(parent)
        } else {
            getUp(parent, levels - 1, env)
        }
    }

    /**
     * Navigate to the root ancestor (top-level note with no parent).
     * If this note has no parent, returns itself.
     */
    private fun getRoot(noteVal: NoteVal, env: Environment): NoteVal {
        val parent = env.getParentNote(noteVal.note) ?: return noteVal
        return getRoot(NoteVal(parent), env)
    }
}
